Adjust window SDK-2243

tmoopenn opened the pull request:
- updated documentation to reflect default behavior of writing csv both for webcam and video options. 

- modified text on frames to have black outlines (measurements have been left the same for reference, emotion metrics and expression metrics are in different fonts for comparison for feedback on which font options people think work best. ). Size of text also increased slightly. 

-  cli arguments for resolution changed to match frame-detector. Default resolution of 1280 x 720.  

- help documentation updated 

- printing of metrics to console changed to generally match frame-detector (differences are that frame-detector outputs number of faces it seems, affvisionpy-sample prints out frame number).

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -86,18 +106,22 @@ def get_command_line_parameters(args):
     output_file = args.output
     csv_file = args.file
-    return input_file, data, max_num_of_faces, csv_file, output_file
+    frame_width = int(args.res[0])
```
Would it be better to have these arguments named rather than refer to them by location?

tmoopenn replied:
It would be! I haven't found an easy way with argparse to set variable names for a cli argument with multiple parameters. If anyone knows, let me know! Otherwise, I could just create two constants.

ForestJay replied:
https://stackoverflow.com/questions/33638475/named-argument-in-argparse . I think we also have some good examples of this in BitBucket webops-automation .

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -357,16 +402,18 @@ def write_metrics(frame):
 
-def run(csv_data):
-    """Starting point of the program, initializes the detctor, processes a frame and then writes metrics to frame
+"""Starting point of the program, initializes the detctor, processes a frame and then writes metrics to frame
```
detector is spelled wrong (and was spelled wrong prior to your change).

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -399,7 +460,7 @@ def run(csv_data):
             else:
                 timestamp = int(captureFile.get(cv2.CAP_PROP_POS_MSEC))
-            print("Frame " + str(count) + " timestamp:" + str(round(timestamp, 0)))
+            time_metrics_dict['timestamp'].put(timestamp)
```
So, we no longer print the timestamp? Also, I think it's better to use update.

tmoopenn replied:
We're still printing it but I'm storing it in a queue here so it can be printed later with the other metrics that are getting printed in the callback function results_updated. This adjustment was made to print metrics on a single line to match the console output of frame-detector demo.

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -481,61 +546,73 @@ def draw_affectiva_logo(frame, width, height):
+        current_frame_data["TimeStamp"] = timestamp
+        current_frame_data["faceId"] = fid
+        x0,y0,x1,y1 = get_bounding_box_points(fid)
+        current_frame_data["upperLeftX"],current_frame_data["upperLeftY"], current_frame_data["lowerRightX"],current_frame_data["lowerRightY"] = x0,y0,x1,y1
```
I think there's no easy way to get around setting multiple values in 589 within one line of code but I think that line 590 should be split into 4 lines, ie:
current_frame_data["upperLeftX"] = x0
current_frame_data["upperLeftY"] = y0
current_frame_data["lowerRightX"] = x1
current_frame_data["lowerRightY"] = y1

Also, I'm not keen on these variable names. I think, for example, x0 would be more readable/maintainable as upper_left_x .

tmoopenn replied:
I agree. Is there a preference for snake_case vs camelCase? Currently, I've renamed x0 to upperLeftX for example.

ForestJay replied:
For Python code please stick to PEP-8 : https://www.python.org/dev/peps/pep-0008/#naming-conventions

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -481,61 +546,73 @@ def draw_affectiva_logo(frame, width, height):
+
+def parse_command_line():
+    parser = argparse.ArgumentParser(description="Sample code for demoing affdexface on webcam or a saved video file.\n \
+    By default, the program will run with the camera parameter displaying frames of size 1280 x 720.\n \
```
I think 614 and 615 should be indented further right.

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -547,38 +624,39 @@ def parse_command_line():
                         help="name of the output video file")
     parser.add_argument("-f", "--file", dest="file", required=False, default="default",
                         help="name of the output csv file")
+    parser.add_argument("-r", "--resolution",dest='res',metavar=('width','height'),nargs=2,default=[1280,720], help="resolution in pixels (2-values): width height")
```
This would be more readable to us old folks to have a space after each comma.

ForestJay commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -547,38 +624,39 @@ def parse_command_line():
-        writer = csv.writer(c_file, quoting=csv.QUOTE_ALL)
-        writer.writerows([header_row])
+        keys = csv_data[0].keys()
+        writer = csv.DictWriter(c_file,fieldnames=header_row)
```
I'd add a space after this comma too. I may have missed other lines like this, I should replace my glasses.

tmoopenn replied:
I'll go through and double check for other instances!

tmoopenn commented on README.md:
```diff
@@ -4,62 +4,75 @@ Steps to install the app:
+
+            python3 affvisionpy-sample.py -d <path/to/data/directory> -n <num_of_faces_to_detect> -v </path/to/video/file>
+
+            A csv file is written with the metrics. By default the csv file will be named "default.csv".
```
@ForestJay As in the name of the file should be in upper case?

ForestJay replied:
I mean the abbreviation, So instead of the comment saying, 'A csv file is written with the metrics. By default the csv file will be named "default.csv"' I'd have it say 'A CSV file is written with the metrics. By default the CSV file will be named "default.csv".'

tmoopenn replied:
Gotcha, I'll edit it in the code as well as the README.

tmoopenn commented on README.md:
```diff
@@ -4,62 +4,75 @@ Steps to install the app:
           -f FILE, --file FILE
-                        name of the output csv file
-                        
+                        enable this parameter to save the output metrics in a csv file of your choice
```
Got it, added example usages for these cases.

tmoopenn commented on README.md:
```diff
@@ -4,62 +4,75 @@ Steps to install the app:
-    
-5. When a video file is provided as an input, a csv file is written with the metrics. By default the csv file will
-    be named as the name of the video file.
+
```
Good point, after looking at it in a proper editor, it is awkward. When I was editing the README, I didn't realize there was markup. I've created sections, and shifted the notes under the example call to an additional section below.

tmoopenn commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -57,23 +75,25 @@ def results_updated(self, faces, image):
         global capture_last_ts
         capture_fps = 1000.0 / (image.timestamp() - capture_last_ts)
-        print("cfps: " + str(round(capture_fps, 0)))
+        time_metrics_dict['cfps'].put(capture_fps)
```
I was putting 'cfps' in a queue because I was thinking we might need a buffer to store the metrics in before we print them out. But since the print out is happening in the results_updated callback function, I realize the queue isn't necessary. Now I'm just assigning it a value:
time_metrics_dict['cfps'] = capture_fps
If I'm just setting a value, I don't think update would apply here?

ForestJay replied:
I'm not actually sure. I'd defer to a Pythonista for this. Maybe @ramywafa or @caboteria .

tmoopenn commented:
Issues addressed in updated code: 
- added additional example usages to README and beautified it with markup 
- @stevehunnicutt Addressed additional bug SDK-2256 of missing rows of csv data when face is occluded. Now, a timestamp is written for the frame with a row of NaN values when face is occluded. 
- noticed only 9 boxes were getting printed for expression metrics so added extra box. 
- shifted measurements slightly more left since the numbers were overlaying onto the bounding box 
- Addressed other formatting issues (comma spacing, naming conventions, mispellings, indentation) noted in previous PR comments for SDK-2243.

adennie commented on python-sdk-samples/affvisionpy-sample.py:
```diff
@@ -4,51 +4,79 @@
             expressions_dict[face.get_id()].update(face.get_expressions())
             emotions_dict[face.get_id()].update(face.get_emotions())
+            mood_dict[face.get_id()] = Mood(mood=face.get_mood(), confidence=face.get_confidence(), dominant_emotion=face.get_dominant_emotion().dominant_emotion,
+                                            dominant_emotion_confidence=face.get_dominant_emotion().confidence)
```
it seems a little strange to me to be packaging several different things (the mood, the tracker confidence, the dominant emotion, and the dominant emotion confidence) into a class called Mood.  I could see packaging dom emo and dom emo confidence together into a DominantEmotionMetric class (like is done in the C++ API), but I would keep mood and (tracker) confidence separate from that, and from each other.

tmoopenn replied:
I agree. Initially, the tuple started with mood and dominant emotion and then additional fields got added but logically, they don't fit together. I'll restructure the packaging so it logically makes sense.

adennie commented on README.md:
```diff
@@ -1,68 +1,88 @@
+# Sample Python App for Affdex SDK #
 
-Steps to install the app:
+This script is the python-sdk-sample for the frame detector demo.
```
Suggestion: "This script demonstrates how to use Affectiva's affvisionpy module to process frames from either a webcam or a video file.  It supports displaying the input frames on screen, overlaid with metric values, and can also write the metric values to an output CSV file".
